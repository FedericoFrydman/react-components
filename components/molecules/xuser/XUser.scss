@import '../../assets/styles/colors.scss';

.lsux-xuser {
  align-items: center;
  display: flex;
  justify-content: space-between;
}
