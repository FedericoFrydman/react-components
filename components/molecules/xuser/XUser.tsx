import { useState, useEffect, Ref, forwardRef, useCallback } from 'react';
import cn from 'classnames';

import { Text } from '../../atoms/text/Text';
import { Spinner } from '../../atoms/spinner/Spinner';
import { Button } from '../../molecules/button/Button';
import './XUser.scss';

export interface XUserProps extends Omit<React.HTMLProps<HTMLDivElement>, 'css'> {
  /**
   * Domain name to query
   */
  domainName?: string;

  port?: string;

  env?: string;
  /**
   * Reference
   */
  ref?: Ref<HTMLDivElement>;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
}

const XUserComponent = (
  { domainName = 'legalshield.com', env = 'prod', port = '', classNames = [], ...props }: XUserProps,
  ref: XUserProps['ref'],
) => {
  const [userState, setUserState] = useState({ data: null, error: false, loading: true });

  const fetchData = useCallback(async () => {
    setUserState({ data: undefined, error: false, loading: true });

    const envString = env === 'prod' ? '' : env + '-';
    const portString = port === '' ? '' : ':' + port;

    const url = `https://login.${envString}${domainName}${portString}/v1/users/info`;

    try {
      const response = await fetch(url, {
        credentials: 'include',
        headers: {
          'Content-Type': 'application/json',
        },
        method: 'GET',
      });

      if (response.ok && response.status >= 200 && response.status < 300) {
        const responseData = await response.json();
        setUserState({ data: responseData, error: false, loading: false });
        return;
      } else {
        console.error(response);
      }
    } catch {}

    setUserState({ data: null, error: true, loading: false });
  }, [domainName, env, port]);

  useEffect(() => {
    window.addEventListener('focus', fetchData);
    // Specify how to clean up after this effect:
    return () => {
      window.removeEventListener('focus', fetchData);
    };
  }, [fetchData]);

  // Load the data
  useEffect(() => {
    fetchData();
  }, [fetchData]);

  let current = 'User in ' + domainName + ': ';
  if (userState.loading) {
    current = current + 'Loading...';
  } else if (userState.error) {
    current = current + 'Error Fetching';
  } else if (!('sub_name' in userState.data)) {
    current = current + 'None';
  } else {
    current = current + userState.data.sub_name + ' (' + userState.data.sub + ')';
  }

  return (
    <div {...props} className={cn('lsux-xuser', classNames)} ref={ref}>
      <Text text={current} as="p" />
      {userState.loading && <Spinner spinnerSize="small" />}
      {!userState.loading && (
        <Button
          variant="tertiary"
          onClick={() => {
            fetchData();
          }}
          iconLeft="action_refresh"
          label="Refresh"
          classNames={['p-0']}
        />
      )}
    </div>
  );
};

export const XUser = forwardRef<HTMLDivElement, XUserProps>(XUserComponent);
