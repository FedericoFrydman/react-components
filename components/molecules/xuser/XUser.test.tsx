import { createRef } from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { render, screen } from '@testing-library/react';
import cn from 'classnames';

import { XUser, XUserProps } from './XUser';

describe('XUser Component', () => {
  let testId: string;
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<XUser />);
    testId = 'test-xuser';
  });

  it('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the XUser the provided classNames', () => {
          const XUser = renderXUser({ classNames });

          expect(XUser).toHaveClass(cn(classNames), { exact: false });
        });
      });
    });
  });

  describe('ref', () => {
    const ref = createRef<HTMLDivElement>();

    it('provides a ref to the xuser', () => {
      expect(renderXUser({ ref })).toEqual(ref.current);
    });
  });

  const renderXUser = (props: XUserProps = {}): HTMLElement => {
    render(<XUser {...props} data-testid={testId} />);

    return screen.getByTestId(testId);
  };
});
