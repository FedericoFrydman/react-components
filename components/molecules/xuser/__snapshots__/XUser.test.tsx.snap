// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XUser Component should render the component 1`] = `
<div
  className="lsux-xuser"
>
  <ForwardRef(TextComponent)
    as="p"
    text="User in legalshield.com: Loading..."
  />
  <ForwardRef(SpinnerComponent)
    spinnerSize="small"
  />
</div>
`;
